Gas Reaction Datums

duncathan opened the pull request:
Finally modularizes reactions out of the awful `/datum/gas_mixture/proc/react()` that they were before. Now new reactions can be added intuitively by adding a new subtype of `/datum/gas_reaction`. 

They can interact with a tile, set requirements to actually react, return either a failure or a success, and modify a new `reaction_results` list on the gas mixture to have results that are accessible until the next tick of `react()`. This last point was an expansion of the `fuel_burnt` var gas mixtures already had. 

Finally, you can designate a particular reaction to be excluded from the reactions singleton so that it will never occur (ex: fusion). In the future I intend to let admins toy around and add or remove reactions at will, so they could theoretically enable fusion for a round should they want to.

duncathan committed:
adds gas reaction datums, converts existing reactions to the new system

duncathan committed:
slightly reorganizes gas_mixture.dm

duncathan committed:
fixes compilation errors

Cyberboss commented:
# H O L Y S H I T

Cyberboss commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -0,0 +1,209 @@
+var/list/gas_reactions = init_gas_reactions() //this is our singleton of all reactions
+/proc/init_gas_reactions()
+	. = new /list
+	var/list/reaction_types = typesof(/datum/gas_reaction - /datum/gas_reaction)
```
use `subtypesof(/datum/gas_reaction)`

duncathan replied:
lol I knew I was rusty

Iamgoofball commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -0,0 +1,209 @@
+/datum/gas_reaction/freon/react(datum/gas_mixture/air, turf/open/location)
+	. = NO_REACTION
+	if(location.freon_gas_act())
+		air.gases["freon"][MOLES] -= MOLES_PLASMA_VISIBLE
```
please remove this I keep meaning to

Cyberboss replied:
We tried already remember? #24731

Iamgoofball replied:
fuckig hell I added the decay because scrubbers didn't work at the time

duncathan replied:
nah, it can be removed in a later PR but as always I want to maintain functionality as much as possible

Jalleo replied:
Not to mention feature freeze.

PKPenguin321 commented:
duncc is it really you

AnturK commented:
Nice, much more readable.

Shadowlight213 commented:
What does this do?

Razharas reviewed (approved):
Finally
A good atmos pr for a change
Take notes people

Razharas commented on code/game/turfs/open.dm:
```diff
@@ -119,7 +117,7 @@
 	MakeSlippery(TURF_WET_PERMAFROST, 5)
 	return 1
```
This whole proc is so bad its painful to watch
1 - make the effect proc universal for atoms
2 - move all checks inside their appropriate object procs
3 - loop once over things in contents and call the effect proc of those movables
4 - point at goof and laugh

duncathan replied:
better idea: ignore this and then just remove freon after the freeze is over

Iamgoofball replied:
Remove freon and I remove you

duncathan replied:
You code in my universe, normie.

Iamgoofball replied:
im the king

you cant dethrone me

duncathan replied:
![You code in my universe, normie.](http://puu.sh/v8lLo/99200d3420.png)

Iamgoofball replied:
Fuck you leatherhead

optimumtact replied:
FREON club is three blocks down

Razharas commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -0,0 +1,209 @@
+	. = new /list
+	var/list/reaction_types = typesof(/datum/gas_reaction - /datum/gas_reaction)
+
+	for(var/i in 2 to reaction_types.len) //holy shit there's gotta be a better way of sorting this lmao. sorts in descending order of priority
```
This is an incomplete version of a poor sorting algorythm

optimumtact replied:
should really just use our timsort algo

duncathan replied:
haven't pushed yet but that is what i ended up doing

Razharas commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -313,6 +170,10 @@ var/list/gaslist_cache = init_gaslist_cache()
 	//Compares sample to self to see if within acceptable ranges that group processing may be enabled
 	//Returns: a string indicating what check failed, or "" if check passes
 
+/datum/gas_mixture/proc/react(turf/open/dump_location)
```
What is this for if you redefine this one 300 lines below

duncathan replied:
This is how all these procs are defined in `gas_mixture.dm`. Defined empty and documented, and then after all the definitions they're given content. Apparently doing this actually​ marginally increases performance, and also makes it very easy to document them.

Cyberboss replied:
>Increases performance?

Please do this for `/turf/process_cell`

optimumtact replied:
@duncathan proof for the increases performance, I'm not a fan of this, it's needlessly confusing and causes grepping for proc paths to be extra noisy.

duncathan replied:
!(Profile)[http://puu.sh/v8DPb/f492765f23.png]
```DM
/datum/profile
	var/iterations = 1e4

	proc/profile1(a)
		return a++

	proc/profile2(a)
	profile2(a)
		return a++

client/verb/profile()
	set background = 1

	var/datum/profile/p = new
	for(var/i in 1 to p.iterations)
		world << "[(i/p.iterations)*100]% complete."
		. += p.profile1(.)
		. += p.profile2(.)
```

optimumtact replied:
what the fuck

optimumtact replied:
@tgstation/commit-access @MrStonedOne what the actual fuck

duncathan replied:
breaking news: byond is stupid

Jordie0608 replied:
This is cool although at most I'd suggest you move this line closer to where you have the actual code for clarity.

duncathan replied:
Look at it in context of the file. Each proc is defined and documented. Once they've all been, it goes through in the order they were defined and gives them a proper definition. It's consistent with the file and makes for clearer documentation.

Jordie0608 replied:
Both are valid ways of file structuring, I just prefer the other. It does't really matter anyway.

AnturK replied:
Yeah, this is pretty insane if this really has any bearing on performance.

duncathan replied:
It's a greater impact than I thought, too. The overhead from defining the first way is 4 times as much as the second way... actually ridiculous.

AnturK replied:
Yeah, if anything i'd expect difference on the scale of linear list access not this.

MrStonedOne replied:
https://pastebin.com/YBqG4Etv

Play with this version.

Iamgoofball replied:
what the fuck is that relative pathing get the fuck out mso

xxalpha replied:
Has this been tested with the new method and without? Is it possible the new method only appears faster because calling the original proc is somehow more expensive only when there are overrides?

coiax replied:
Instead of doing this weird fucking pathing shit, can't we just do this automatically with some sort of pre-compile script?

MrStonedOne replied:
\>nobody actually played with my version to profile it

protip: the difference basically goes away

its the last call bullshit in byond

Razharas replied:
Since its a proven bullshit you dont need to do this dunc

Iamgoofball replied:
But Cyberboss just did a test proving it worked

duncathan replied:
well I don't need to define it like this but I will anyway because it makes documentation so much better for a file that very much needs it.

xxalpha replied:
He didn't use mso's test, which randomizes the proc calls. What MSO has discovered is that the proc called second in these types of tests will be called (or executed?) faster, so if you swap the proc call order you get different results, so it's not a reliable test.

duncathan committed:
no need to roll my own code when procs already exist

duncathan committed:
Merge remote-tracking branch 'upstream/master' into reactiondatums

duncathan committed:
debug improvements

duncathan committed:
makes gas_reactions initialize properly


apparently arglist doesn't work in newlist so i'll just do it manually


jesus i'm bad

duncathan committed:
bug fixes and optimizations

AnturK commented:
Probably put that debug info behind #ifdef TESTING, saves on performance during normal operation

duncathan commented:
i'm just gonna comment it out actually i forgot to

duncathan commented:
https://pastebin.com/mGuRdHAr
https://pastebin.com/c0mnFvKF

duncathan committed:
minor tweaks: optimizations, cleanup, modular results var

Kromgar commented:
So why is everyone freakign out is this some kind of sorcery that fixes lag?

optimumtact commented:
no it was just unexpected  and proved to be unfounded

duncathan commented:
>[23:39] (Dunc|Mobile) Cheridan: can #25771 be merged? the only thing holding it back is the weirdness with the double define but until we figure it out i don't want my pr held up because i made a change to improve consistency within a file
>...
>[23:42] (@Cheridan) sure why not

Merge please
